### PR TITLE
Bugfix

### PR DESCRIPTION
**Pull request by Franck78**
Ok,
I spent a lot lot of time tracking this bug https://github.com/opnsense/core/issues/2774
Gone for good, I can load then restart any number of time the daemon.

To test, on two terminals:

T1:
#/usr/local/sbin/dhcp6c -f -c /var/etc/dhcp6c_opt1.conf -p /var/run/xxx.pid em0_vlan832
OK
OK
OK

T2:
#kill -HUP `ps axc | grep dhcp6c | awk '{print $1}'`

Franck

### CODE REVIEW

File: common.c [hunk 1]
```diff
@@ -1974,15 +1971,15 @@ dhcp6_get_options(p, ep, optinfo)
 
 			break;
 
-		/* XXX */
+/* FRK  XXX   There is such thing here. RAW is a syntax, not an option !
```
Comment (fichtner) [inline common.c:1974]: if you are sure please drop this from code :)

File: common.c [hunk 2]
```diff
@@ -3172,9 +3169,10 @@ dhcp6optstr(type)
 		return ("subscriber ID");
 	case DH6OPT_CLIENT_FQDN:
 		return ("client FQDN");
-	/* XXX */
+/*	Either a known or an unknown option. RAW is a syntax, not an option
```
Comment (fichtner) [inline common.c:3172]: same here

**Review by fichtner — APPROVED** [review]
ok for me to merge and do a public call for testing

maybe if this goes well we should remove the XXX annotations as we're past the point of having this in a non-ready state.

@Franck78 what do you think? :)

**Comment by Franck78** [comment]
Hello Franco,

marjohn56 wrote somewhere :
"It think I'll make it my summer project for...when"
about Dibbler. If this is the case, then suppress the XXX is ok.

If Dibbler is not for now then :

the struct "struct rawoption" should not exist ;)
-no need for opnum field alone. The opnum+bytes are a single block of data
raw-option 11 41:42:43 is 0B:41:42:43 in the end.
And the struct with a variable length buffer is implemented :
struct dhcp6_vbuf {     /* generic variable length buffer */
0116     int dv_len;
0117     caddr_t dv_buf;
0118 };

which mean basically adding to the enum  dhcp6_listval_type_t
a raw_option

Second what when a 'rawoption' is really a known option ?
11 41:42:43
DH6OPT_AUTH  "ABC"
same thing. But not same verifications/handling.

**Comment by marjohn56** [comment]
I will still look at dibbler, as @nivek1612 and I did have it working under pf - all that code is now lost so I would need to start afresh and get my head around it again. I will also contact the author and see if support is ongoing, otherwise we are exchanging one unsupported application for another. Dibbler is better overall though even then -  as it does support a lot more features than dhcp6c.
In the meantime its probably a good idea to go with the changes @Franck78 has done and give it some testing.

**Comment by marjohn56** [comment]
Quick note, looks like Tomas is sill supporting Dibbler as a PR in January was pulled. I've just re-forked it and compiled it with no issues so I'll play around with it and see where it takes me.

**Comment by marjohn56** [comment]
@fichtner - pull this in for 19.7 and lets give it a thorough test. I have looked deeper into Dibbler and although I have it working it requires scripts to apply the received addresses and ranges to the interfaces as it does not directly support those actions under FreeBSD, dhcp6c in that sense is far better. I suggest we keep improving dhcp6c the way we are doing.

**Comment by Franck78** [comment]
Ok then, i will remove the XXX after I tried to 'remove' the extra struct rawoption.

**Comment by fichtner** [comment]
@Franck78 @marjohn56 thanks guys, ready for pull when you are

**Comment by nivek1612** [comment]
I land in France later today so can also stress test it in the live if it’s ready

**Comment by Franck78** [comment]
I don't know when I will have time to do what I want. But it is not needed for validating the bug fix.

Also, anyone can try to trigger the bug. For that, you need to add some rawoption lines (with a code that will be ignored by the server of course).
Then restart dhcp6c with dashboard
Until you have a rawoption set that messes up the config.

**Comment by fichtner** [comment]
Then I'll merge and remove the XXX for the next release. Thank you @Franck78, very nice work!

**Comment by marjohn56** [comment]
@nivek1612 I'll clone the source from the repo later , compile it and send you the exe.

**Comment by fichtner** [comment]
# pkg add -f https://pkg.opnsense.org/FreeBSD:11:amd64/snapshots/libressl/All/dhcp6c-20190419.txz

**Comment by marjohn56** [comment]
Or do it that way 😀

**Comment by nivek1612** [comment]
That’s too easy I’ll compile it from the source ..... then once I screw it up I’ll do it that way :-)

**Comment by fichtner** [comment]
Could also send a fax with the current source and you have to type it before compiling. Good code review :)

**Comment by nivek1612** [comment]
LOL

I'm going the easy way 

just in case there are any issues whats the command to reverse the #pkg add ?

**Comment by fichtner** [comment]
# opnsense-revert dhcp6c

**Comment by nivek1612** [comment]
Well I can't break it so looks like its good to go. Tried reboots, save wan, drop wan, stopped dhcp6c, restarted it. All seems good

**Comment by fichtner** [comment]
@nivek1612 neat, we'll have this in 19.1.7 then. thanks everyone 👍

**Comment by marjohn56** [comment]
I suspect this all came about when I modified it to reload the config on a SIGHUP, previously to reload the config required a restart, so kudos to @Franck78 for fixing it.

**Comment by Franck78** [comment]
Hi,
nothing to do with restart. I had a config that make it crashes at first start. The real bug is in config.c, ~line 1700, the object is copied partially from a list to another. Then the object is freed. Entirely. But the copy object still have pointer that are now invalid. So the apparent  random readings of parameters. 
The rest of the patch is ~cosmetic .

**Comment by marjohn56** [comment]
Still bizarre that I could never trigger the failure though.

**Comment by fichtner** [comment]
No worries. Progress is progress we like to say.  😊👍

**Comment by Franck78** [comment]
Memory allocation stuf!

And about that, the code is not easy to read because of the numerous "p=malloc(), if (p == NULL)" error blocks. Maybe 15 years ago that was OK to check on a 256Mo memory system.


But now every hardware have tons of gbyte , block sizes are still ~64 bytes. If the system returns NULL, well opnsense is already dead.

So I was thinking (yes again) get rid of all those disseminated  tests and just at startup do a single loop allocating a hundred blocks them freeing them as single check.

**Comment by marjohn56** [comment]
If it ain't broke... don't fix it. 

However, if you've nothing better to do. :)

**Pull request merged**